### PR TITLE
Maven plugin extensions

### PR DESCRIPTION
**Pull request by ishestakov**
At the moment to extend some Carnotzet functionality you have to register a new instance of `CarnotzetExtension` implementation over Java API but there was no way to extend Carnotzet core functionality in Maven builds. This pull request will make it possible.

The idea is to utilize [`java.util.ServiceLoader`](https://docs.oracle.com/javase/6/docs/api/java/util/ServiceLoader.html) for that and define which extensions of Carnotzet you want to use in your `pom.xml`:
```xml
 <plugins>
    <plugin>
        <groupId>com.github.swissquote</groupId>
        <artifactId>zet-maven-plugin</artifactId>
        <version>${project.version}</version>
        <dependencies>
          <dependency>
            <groupId>com.github.swissquote.examples</groupId>
            <artifactId>hello-extension</artifactId>
            <version>${project.version}</version>
            </dependency>
          </dependencies>
      </plugin>
  </plugins>
```
Like this the `hello-extension` artifact will provide concrete implementation of `CarnotzetExtension` and descriptor file `META-INF/services/com.github.swissquote.carnotzet.core.CarnotzetExtension` for it.

Like this it would be possible to extend core functionality of Carnotzet maven plugin.

### CODE REVIEW

**Comment by mryan43** [comment]
what about extensions that are configurable (constructor arg or properties) ?

**Comment by ishestakov** [comment]
Interesting point which didn't came into my mind. At the moment I don't see lots of options except:
1. Provide life-cycle method `CarnotzetExtension#init(Carnotzet)` to configure and initialize an extension
2. Completely change the approach and use some lightweight IoC container (e.g. Pico or Guice)
3. Keep everything as it is at the moment and let the developer of extensions to initialize it in the default constructor or at the `CarnotzetExtension#apply` method call.
IMO, option 2 is cumbersome.

Do you see any other way?

**Comment by mryan43** [comment]
About option 2, maven already has an IoC container (plexus) and a mechanism to configure plugins with XML (https://maven.apache.org/guides/mini/guide-configuring-plugins.html)

I think the best for usage is to allow configuration of the extensions through the configuration of the zet plugin itself. Since we are not directly configuring the plugin, we would probably need some kind of convention. It could look something like this : 

```
<configuration>
    <extensions>
        <extension>
            <className>com.example.MyExtension</className>
            <configuration>
                <option1>value1</option1>
                <option2>value2</option2>
            </configuration>
        </extension>
    </extensions>
</configuration>
```

I'm not yet sure how we can to use maven's configuration object mapper to achieve this though. What do you think ?

**Comment by ishestakov** [comment]
As configuration related only for a particular Maven build I think it make sense to have an extension configuration inside zet plugin itself and not any other resource (e.g. `carnotzet.properties`, etc.)

I investigate on Maven and DI approach it uses. Unfortunately, the idea of Maven is against the goal we try to achieve with runtime components discovery:
>  Note that our container is configured by default to only use the index. While this keeps things fast, if you use JSR-330 components in dependencies that do not contain an index, those implementations will currently not be discovered. This is a compromise that is reasonable given Maven is a command-line tool where startup speed is important.
[http://maven.apache.org/maven-jsr330.html](url)

Taking all above into account your proposal looks like the best solution at the moment.
I made a change so now the configuration could be done with a slight change.

Please, take a look on it.

**Comment by mryan43** [comment]
Difference between carnotzet.properties and CarnotzetConfiguration : 

- carnotzet.properties contains "module specific" configuration, that affects behavior in ALL environments where the module is present.

- CarnotzetConfiguration is not transitive and only affects one environment.

Extension configuration may use the same approach for the same needs.

I'm currently reviewing the changes.

File: e2e-tests/src/test/java/com/swissquote/github/carnotzet/e2e/test/ExtensionExamplesTest.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+	@Test
+	public void testClassPathExtension() throws VerificationException, IOException {
+
+		mavenVerifier.verify(false);
+		mavenVerifier.executeGoal("zet:start");
+
+		String output = CommandRunner
+				.runCommandAndCaptureOutput("docker", "ps", "--filter", "label=carnotzet.hello.message=Hello Carnotzet");
+		Assert.assertThat(output.split("\n").length, CoreMatchers.equalTo(4));
```
Comment (mryan43) [inline e2e-tests/src/test/java/com/swissquote/github/carnotzet/e2e/test/ExtensionExamplesTest.java]: equalTo(4) is a bit cryptic here.

You can add -q to the docker command and assert that the output is not empty.

File: e2e-tests/src/test/java/com/swissquote/github/carnotzet/e2e/test/ExtensionExamplesTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import com.github.swissquote.carnotzet.core.runtime.CommandRunner;
+
+public class ExtensionExamplesTest {
+
+	private static File logFile;
+	private static Verifier mavenVerifier;
+
+	@BeforeClass
+	public static void setUp() throws IOException, VerificationException {
+		Path projectPom = FileSystems.getDefault().getPath("../examples/voting-worker").toRealPath();
```
Comment (mryan43) [inline e2e-tests/src/test/java/com/swissquote/github/carnotzet/e2e/test/ExtensionExamplesTest.java]: Please run redis example module instead to reduce test execution time. (you'll have to move the maven plugin config)

File: maven-plugin/src/main/java/com/github/swissquote/carnotzet/maven/plugin/AbstractZetMojo.java [hunk 3]
```diff
@@ -1,30 +1,36 @@
...
+import org.apache.maven.settings.Settings;
+import org.slf4j.bridge.SLF4JBridgeHandler;
+
+import java.nio.file.Paths;
+import java.util.Properties;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.Optional;
+import java.util.stream.Collectors;
```
Comment (mryan43) [inline maven-plugin/src/main/java/com/github/swissquote/carnotzet/maven/plugin/AbstractZetMojo.java]: Please use the eclipse/intellij code style provided in code_style if you want to use your IDE import management option. (we should add some checkstyle rules)

**Pull request merged**